tap_appsflyer/streams/abstracts.py

```diff
@@ -0,0 +1,357 @@
+    next_page_key = "next_page"
+    headers = {"Accept": "application/json"}
+
+    def __init__(self, client=None) -> None:
+        self.client = client
+        self.params = {}
+
+    @property
+    @abstractmethod
+    def tap_stream_id(self) -> str:
+        """Unique identifier for the stream.
+
+        This is allowed to be different from the name of the stream, in
+        order to allow for sources that have duplicate stream names.
+        """
+
+    @property
+    @abstractmethod
+    def replication_method(self) -> str:
+        """Defines the sync mode of a stream."""
+
+    @property
+    @abstractmethod
+    def replication_keys(self) -> str:
+        """Defines the replication key for incremental sync mode of a
+        stream."""
+
+    @property
+    @abstractmethod
+    def forced_replication_method(self) -> str:
+        """Defines the sync mode of a stream."""
+
+    @property
+    @abstractmethod
+    def key_properties(self) -> Tuple[str, str]:
+        """List of key properties for stream."""
+
+    @property
+    def selected_by_default(self) -> bool:
+        """Indicates if a node in the schema should be replicated, if a user
+        has not expressed any opinion on whether or not to replicate it."""
+        return False
+
+    def get_url_endpoint(self) -> str:
+        """
+        Get the URL endpoint for the stream
+        """
+        return self.url_endpoint
+
+    @abstractmethod
+    def sync(
+        self, state: Dict, schema: Dict, stream_metadata: Dict, transformer: Transformer
+    ) -> Dict:
+        """
+        Performs a replication sync for the stream.
+        ~~~
+        Args:
+         - state (dict): represents the state file for the tap.
+         - schema (dict): Schema of the stream
+         - transformer (object): A Object of the singer.transformer class.
+
+        Returns:
+         - bool: The return value. True for success, False otherwise.
+
+        Docs:
+         - https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md
+        """
+
+    def parse_source_from_url(self, base_url):
+        url_regex = re.compile(base_url + r".*/(\w+)_report/v5")
+        match = url_regex.match(base_url)
+        if match:
+            return match.group(1)
+        return None
+
+    def get_records(self) -> List:
+        """Interacts with api client interaction and pagination."""
+        extraction_url = self.url_endpoint
+
+        response = self.client.get(extraction_url, self.params, self.headers)
+        if not response:
+            LOGGER.warning("No records found in the response")
+
+        with singer.metrics.http_request_timer(
+            self.parse_source_from_url(self.client.base_url)
+        ) as timer:
+            resp = SESSION.send(response)
+            timer.tags[singer.metrics.Tag.http_status_code] = resp.status_code
+        return resp
+
+    def write_schema(self, schema, stream_name):
+        """
+        Write a schema message.
+        """
```

Comment:
Be consistent with comment pattern.

```suggestion
        """Write a schema message."""
```